src/components/Scaler/index.js

```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
-        xDomain: prevXDomain,
-        externalXDomain: prevExternalXDomain,
-      },
+      dataContext: { timeDomain: prevTimeDomain },
     } = prevProps;
-    if (!isEqual(prevExternalXDomain, nextExternalXDomain)) {
-      // External base domain changed (props on DataProvider)
-      // Reset state
-      // eslint-disable-next-line react/no-did-update-set-state
-      this.setState({
-        xSubDomain: nextExternalXDomain,
-        ySubDomains: {},
-        yTransformations: {},
-      });
-      return;
-    }
+
     if (
-      nextPropsDomain[0] !== prevXDomain[0] ||
-      nextPropsDomain[1] !== prevXDomain[1]
+      nextTimeDomain[0] !== prevTimeDomain[0] ||
+      nextTimeDomain[1] !== prevTimeDomain[1]
     ) {
-      // The internal xDomain changed
-      // Keep existing subdomain
-      const { xSubDomain } = prevState;
-      if (
-        xSubDomain &&
-        (xSubDomain[1] === prevXDomain[1] || xSubDomain[1] >= prevXDomain[1])
-      ) {
-        // You are looking at the end of the window
-        // and the xDomain is updated
-        // Lock the xSubDomain to the end of the window
-        const dt = xSubDomain[1] - xSubDomain[0];
-        // eslint-disable-next-line react/no-did-update-set-state
-        this.setState({
-          xSubDomain: [nextPropsDomain[1] - dt, nextPropsDomain[1]],
-        });
-      }
-      if (
-        xSubDomain &&
-        (xSubDomain[0] === prevXDomain[0] ||
-          xSubDomain[0] <= nextPropsDomain[0])
-      ) {
-        // You are looking at the front of the window
-        // and the base domain is updated.
-        // Lock the sub domain to the start of the window
-        const dt = xSubDomain[1] - xSubDomain[0];
-        // eslint-disable-next-line react/no-did-update-set-state
-        this.setState({
-          xSubDomain: [nextPropsDomain[0], nextPropsDomain[0] + dt],
+      const subDomainsByItemId = {};
+      []
+        .concat(this.props.dataContext.series)
+        .concat(this.props.dataContext.collections)
+        .forEach(item => {
+          const { [Axes.time]: timeSubDomain } = this.state.subDomainsByItemId[
+            item.id
+          ];
+          subDomainsByItemId[item.id] = {
+            ...this.state.subDomainsByItemId[item.id],
+          };
+          const dt = timeSubDomain[1] - timeSubDomain[0];
+          if (
+            Math.abs((timeSubDomain[1] - prevTimeDomain[1]) / dt) <=
+            FRONT_OF_WINDOW_THRESHOLD
+          ) {
+            // Looking at the front of the window -- continue to track that.
+            subDomainsByItemId[item.id][Axes.time] = [
+              nextTimeDomain[1] - dt,
+              nextTimeDomain[1],
+            ];
+          } else if (timeSubDomain[0] <= prevTimeDomain[0]) {
+            // Looking at the back of the window -- continue to track that.
+            subDomainsByItemId[item.id][Axes.time] = [
+              prevTimeDomain[0],
+              prevTimeDomain[0] + dt,
+            ];
+          }
```

Comment:
I don't see the need to completely reconstruct `subDomainsByItemId` here. Couldn't you just iterate through the current state's  `subDomainsByItemId` and change what needs to be changed (like what you did above when the previous timeSubDomain was not equal to the new one)?

Reply:
Yeah they're probably the same. (I was just copying the window-tracking logic from InteractionLayer over without following what it did)